### PR TITLE
write body for run_knob_identification

### PR DESCRIPTION
**Pull request by swimj**
I found and used the preprocessing.Standardize class as it looked it was doing standardization. If I am wrong, let me know and I'll adjust it to use sklearn's StandardScaler.  In that case, should I make a class that does this and put it in preprocessing.py?

### CODE REVIEW

File: server/website/website/tasks/periodic_tasks.py [hunk 1]
```diff
@@ -9,14 +9,14 @@
...
 from analysis.lasso import LassoPath
-from analysis.preprocessing import Bin, get_shuffle_indices
+from analysis.preprocessing import Bin, get_shuffle_indices, Standardize
 
 from website.models import PipelineData, PipelineRun, Result, Workload
 from website.types import PipelineTaskType
 from website.utils import DataUtil, JSONUtil
 
 
-# Log debug messages
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 2]
```diff
@@ -25,15 +25,15 @@
...
-#Note: This annotation supports more time units (minute/hour/day). The 
+#Note: This annotation supports more time units (minute/hour/day). The
 #following setting tells celery to run background job every 1 hour.
 
 #@periodic_task(run_every=(crontab(hour=1,minute=0,day_of_week=0)),name="run_background_tasks")
 
 def run_background_tasks():
     ## 1. Create a new entry in the PipelineRun table for this
     ##    background task
-
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def run_background_tasks():
     for workload in unique_workloads:
         ## 2a. Call first subtask to aggregates the knob & metric data for
         ##     this workload. Save the knob & metric data as (separate)
-        ##     entries in the PipelineData table
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 4]
```diff
@@ -156,7 +156,7 @@ def aggregate_data(workload):
     metric_data = {
         'data': aggregated_data['y_matrix'],
         'rowlabels': copy.deepcopy(aggregated_data['rowlabels']),
-        'columnlabels': aggregated_data['y_columnlabels']
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 5]
```diff
@@ -187,7 +187,7 @@ def run_workload_characterization(metric_data):
             nonconst_columnlabels.append(cl)
     nonconst_matrix = np.hstack(nonconst_matrix)
     n_rows, n_cols = nonconst_matrix.shape
-
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 6]
```diff
@@ -214,10 +214,10 @@ def run_workload_characterization(metric_data):
     detk = create_kselection_model("det-k")
     detk.fit(components, kmeans_models.cluster_map_)
 
-    # Get pruned metrics, cloest samples of each cluster center
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Delete space at end of line

File: server/website/website/tasks/periodic_tasks.py [hunk 7]
```diff
@@ -236,41 +236,42 @@ def run_knob_identification(knob_data, metric_data):
...
+    knob_columnlabels = knob_data['columnlabels']
+
+    metric_matrix = metric_data['data']
+    metric_rowlabels = metric_data['rowlabels']
+    metric_columnlabels = metric_data['columnlabels']
+
+    nonconst_knob_matrix = []
+    nonconst_knob_columnlabels = []
+
+    for col, cl in zip(knob_matrix.T, knob_columnlabels):
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Preprocessing comment

File: server/website/website/tasks/periodic_tasks.py [hunk 8]
```diff
@@ -236,41 +236,42 @@ def run_knob_identification(knob_data, metric_data):
...
+        if np.any(col != col[0]):
+            nonconst_knob_matrix.append(col.reshape(-1, 1))
+            nonconst_knob_columnlabels.append(cl)
+    nonconst_knob_matrix = np.hstack(nonconst_knob_matrix)
+
+    nonconst_metric_matrix = []
+    nonconst_metric_columnlabels = []
+
+    for col, cl in zip(knob_matrix.T, metric_columnlabels):
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Preprocessing comment

File: server/website/website/tasks/periodic_tasks.py [hunk 9]
```diff
@@ -236,41 +236,42 @@ def run_knob_identification(knob_data, metric_data):
...
+        if np.any(col != col[0]):
+            nonconst_metric_matrix.append(col.reshape(-1, 1))
+            nonconst_metric_columnlabels.append(cl)
+    nonconst_metric_matrix = np.hstack(nonconst_metric_matrix)
+
+    standardizer = Standardize(axis = 0)
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Preprocessing comment

File: server/website/website/tasks/periodic_tasks.py [hunk 10]
```diff
@@ -236,41 +236,42 @@ def run_knob_identification(knob_data, metric_data):
...
+    standardized_knob_matrix = standardizer.fit_transform(nonconst_knob_matrix)
+    standardized_metric_matrix = standardizer.fit_transform(nonconst_metric_matrix)
+
+    shuffle_indices = get_shuffle_indices(standardized_knob_matrix.shape[0], seed=17)
```
Comment (dvanaken) [inline server/website/website/tasks/periodic_tasks.py]: Preprocessing comment

**Comment by swimj** [comment]
Ok, I switched out the scaling to use StandardScaler and added some comments.

Regarding the whitespace, I'm not exactly sure what is causing that.  And I think what actually happened is that there were spaces at the end of some lines in the original code, and my editor (Atom) must have gotten rid of them by default. Since when I look at the diff between my commit and the original, the lines highlighted red are the ones with the extra spaces. Do you mean I should go back and put the spaces in? 

In the future, I think there is a setting I can turn off so that trailing spaces in any line do not get trimmed. Should I do that?

**Comment by dvanaken** [comment]
@swimj LGTM. Once you update your branch with the new changes I will merge.

**Comment by swimj** [comment]
Sorry, I'm still new to this. What do I do to update my branch?

**Comment by dvanaken** [comment]
No worries! I was actually able to merge it in manually.

**Pull request merged**